arch: arc: implement DIRECT IRQ support

vonhust opened the pull request:
* implement DIRECT IRQ support both for normal irq and fast irq.
* add separate interrupt stack for fast irq and use CONFIG_ARC_
  _FIRQ_STACK to control it. This will bring shortest interrupt
  latency for fast irq.
* note that scheduling in DIRECT IRQ is not supported.

Commit:
arch: arc: implement DIRECT IRQ support

* implement DIRECT IRQ support both for normal irq and fast irq.
* add separate interrupt stack for fast irq and use CONFIG_ARC_
  _FIRQ_STACK to control it. This will bring shortest interrupt
  latency for fast irq.
* note that scheduing in DIRECT IRQ is not supported.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

ioannisg commented:
I think this is worth to mention in the 2.1 release notes.

Commit:
tests: add the case for ARC in yaml

* add the case for ARC in yaml after dynamic and direct irq are
  supported
* fix the bug that index in sw_isr_table should have a offset of
  CONFIG_GEN_IRQ_START_VECTOR

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

abrodkin reviewed (changes requested):
Good that we have it finally but a couple of comments need addressing.

Commit:
soc: nsim_hs: fix the interrupt numbers

it's bug found after implementing direct irq support,
nsim_hs has the same number of interrupts available
as nsim_em

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

abrodkin commented on soc/arc/snps_nsim/Kconfig.defconfig.hs:
```diff
@@ -13,7 +13,7 @@ config NUM_IRQ_PRIO_LEVELS
 config NUM_IRQS
 	# must be > the highest interrupt number used
-	default 20
+	default 30
```
1. How this commit is related to "Direct IRQ" pull-request? I'd say it is completely orthogonal so let's move it to a separate pull-request.
2. I guess you meant "Number of interrupts available" but not "interrupt numbers".

vonhust replied:
After implementing direct IRQ support, nsim_hs will fail on compiling for the tests/kernel/gen_isr_table because of wrong setting in NUM_IRQs.  So  I think it shoud be part of this pull request.

abrodkin replied:
That's OK then but in that case it's good to explain that in the commit message so it's immediately clear what are this change relationships with the main feature of the pull-request.

vonhust replied:
done

abrodkin commented:
@vonhust I see you pushed new version of your patches but looks like you just added support for tests but none of my comments were addressed. Are you still going to check my comments?

vonhust commented on include/arch/arc/v2/irq.h:
```diff
@@ -68,6 +69,79 @@ extern void z_irq_spurious(void *unused);
+	if (IS_ENABLED(CONFIG_ARC_FIRQ) && \
+	    !IS_ENABLED(CONFIG_ARC_FIRQ_STACK) && \
+	    priority_p == 0) { \
+		z_irq_priority_set(irq_p, priority_p + 1, flags_p); \
```
ok , 1 is better

abrodkin replied:
Well and maybe in comments explain why we use 1.
Maybe something like that in comments:

There're 2 special cases when we don't want to use ARC FastIRQ (FIRQ) and so instead of requested priority 0 set it to the next - 1.
1. FIRQ doesn't have a separate stack (CONFIG_ARC_FIRQ_STACK = y), it's not safe to call C ISR handlers. *Even though it might be obvious to you today please explain why separate FIRQ stack is needed for C ISR handlers as reader might assume that normal IRQ stack could be used for FIRQs as well*

2. FIRQ has separate stack (CONFIG_ARC_FIRQ_STACK=y) but at the same time stack checking is enabled (CONFIG_ARC_STACK_CHECKING=y). Stack checking logic will raise stack check exception as SP is also banked and so it will be used from bank1.

vonhust commented on arch/arc/core/reset.S:
```diff
@@ -143,6 +144,13 @@ _slave_core_wait:
 	_get_curr_cpu_irq_stack sp
 
+#if defined(CONFIG_ARC_FIRQ_STACK)
+	mov_s r0, _firq_interrupt_stack
```
done

vonhust commented on arch/arc/core/reset.S:
```diff
@@ -167,4 +175,9 @@ _master_core_startup:
 	add sp, sp, INIT_STACK_SIZE
 
+#if defined(CONFIG_ARC_FIRQ_STACK)
+	mov_s r0, _firq_interrupt_stack
```
done

vonhust commented on arch/arc/core/fast_irq.S:
```diff
@@ -22,6 +22,23 @@
+#if defined(CONFIG_ARC_FIRQ_STACK)
+GTEXT(z_arc_firq_stack_set)
+SECTION_FUNC(TEXT, z_arc_firq_stack_set)
+	clri r1
```
the z_arc_firq_stack_set must be called when irq diasbled,  it may be called not only in the init phase but also other places.  So clri r1 and seti r1 are needed.  seti r1 may not enable irq but just restore the original IE status.

abrodkin replied:
Then please explain this in comments so that changes I propose are  not implemented by somebody else later :)

andrewboie replied:
add comments for this

vonhust replied:
comments added

vonhust commented on arch/arc/core/fast_irq.S:
```diff
@@ -22,6 +22,23 @@
+GTEXT(z_arc_firq_stack_set)
+SECTION_FUNC(TEXT, z_arc_firq_stack_set)
+	clri r1
+	mov ilink, r0
```
you are right, only ilink will not be banked. i will add some comments

vonhust commented on include/arch/arc/v2/irq.h:
```diff
@@ -68,6 +69,79 @@ extern void z_irq_spurious(void *unused);
+	     IS_ENABLED(CONFIG_ARC_FIRQ_STACK) && \
+	     IS_ENABLED(CONFIG_ARC_STACK_CHECKING) && \
+	     priority_p == 0) { \
+		z_irq_priority_set(irq_p, priority_p + 1, flags_p); \
```
fixed as above

vonhust commented on include/arch/arc/v2/irq.h:
```diff
@@ -68,6 +69,79 @@ extern void z_irq_spurious(void *unused);
+#define Z_ARCH_IRQ_DIRECT_CONNECT(irq_p, priority_p, isr_p, flags_p) \
+({ \
+	Z_ISR_DECLARE(irq_p, ISR_FLAG_DIRECT, isr_p, NULL); \
+	if (IS_ENABLED(CONFIG_ARC_FIRQ) && \
```
It cannot be merged, e.g. the case  CONFIG_ARC_FIRQ + CONFIG_ARC_FIRQ_STACK + priotiy == 0 and no CONFIG_ARC_STACK_CHECKING,  z_irq_priority_set will be skipped.

abrodkin replied:
So what's wrong here? If we have `CONFIG_ARC_FIRQ=y` && `CONFIG_ARC_FIRQ_STACK=y` && `CONFIG_ARC_STACK_CHECKING=n` then we proceed to "else" branch and set priority to 0. No?

vonhust replied:
I got your idea, done

vonhust commented on include/arch/arc/v2/arcv2_irq_unit.h:
```diff
@@ -86,6 +86,22 @@ void z_arc_v2_irq_unit_int_disable(int irq)
+ */
+
+static ALWAYS_INLINE
+bool z_arc_v2_irq_unit_int_enabled(int irq)
```
Just like other functions z_arch_irq_disable/enable,  functions in irq_manage.c should not directly access aux regs.

abrodkin replied:
Didn't know that. Is it explained somewhere?

vonhust replied:
No explaination , just keep the same style and abstract

abrodkin replied:
Well maybe we're overengineering here. But OK.

vonhust commented on arch/arc/core/fast_irq.S:
```diff
@@ -22,6 +22,23 @@
+	kflag r0
+	mov sp, ilink
+	add sp, sp, CONFIG_ARC_FIRQ_STACK_SIZE
+	lr ilink, [_ARC_V2_STATUS32]
```
I will add more comments

vonhust commented on arch/arc/Kconfig:
```diff
@@ -95,6 +95,21 @@ config ARC_FIRQ
+	default n
+	help
+	  Use separate stack for FIRQ handing. When the fast irq is also a direct
+	  irq, this will get the bet interrupt latency.
```
done

ruuddw commented on arch/arc/core/fast_irq.S:
```diff
@@ -22,6 +22,36 @@
+	mov_s r0, _firq_interrupt_stack
+#ifdef CONFIG_SMP
+	_get_cpu_id r1
+	mpy r1, r1, CONFIG_ARC_FIRQ_STACK_SIZE
```
This assumes there is a hw multiplier configured - probably true for most ARC instances, but doesn't have to be...

vonhust replied:
It's a C arrary operatioon. It's a fast way to do this using mpy instruction. Without it, code will become very complex.  I think when firq is configured and enabled, mpy should be there.

abrodkin replied:
@vonhust it that case it worth checking explicitly if `__ARC_MPY__` is defined by GCC and report user an error due to building for unsupported configuration. Otherwise user might be surprised be the compiler error saying something about unknown opcode or something.

vonhust replied:
@abrodkin  mpy is used when SMP and FIRQ  and FIRQ_STACK are all enabled, for this case, it doesn't make sense that mpy is not there. So  do not want to add __ARC_MPY__ for it to make codes complex.

abrodkin replied:
@vonhust As I just learned there's a bigger problem here! Our GCC only checks supported opcodes when it compiles C! But when only assembler & linker are involved user will only know this instruction is not supported on execution when exception happens. Thus I strongly suggest to add a check for mentioned `__ARC_MPY__ ` definition.

vonhust replied:
@abrodkin OK, I turn z_arc_firq_stack_set into a C function with inline assembly,  don't use mpy instruction explictly.

ruuddw commented on include/arch/arc/v2/irq.h:
```diff
@@ -68,6 +69,82 @@ extern void z_irq_spurious(void *unused);
+		IS_ENABLED(CONFIG_ARC_STACK_CHECKING))) { \
+			z_irq_priority_set(irq_p, 1, flags_p); \
+		} else { \
+			z_irq_priority_set(irq_p, 1, flags_p); \
```
both cases are the same?
More importantly, I don't like this automatic change of priority -> instead, generate an #error that tells the user to set the priority to 1

vonhust replied:
I like the idea to output a error msg, but we can't insert #error here, inside a macro definition and inside if.

andrewboie replied:
use BUILD_ASSERT_MSG()

vonhust replied:
Thanks for your suggestion @andrewboie , already done

ruuddw commented on include/arch/arc/v2/irq.h:
```diff
@@ -68,6 +69,82 @@ extern void z_irq_spurious(void *unused);
+	/* clear SW generated interrupt */
+	if (z_arc_v2_aux_reg_read(_ARC_V2_ICAUSE) ==
+	    z_arc_v2_aux_reg_read(_ARC_V2_AUX_IRQ_HINT)) {
+		z_arc_v2_aux_reg_write(_ARC_V2_AUX_IRQ_HINT, 0);
```
If this is only for SW generated interrupts, why not leave this to the isr implementor?

vonhust replied:
E.g. in zephyr's tests/kernel/gen_isr_table,  SW generated interrupts are used, but in the generic ISR, no ARC specific code is there to clear SW irq request.  Leaving the clear of SW int reqs to arch  will make the ISR more reuseable and more generic. This is also the way we did in _isr_demux.

andrewboie replied:
I think this current code is OK, by the time we get here we have already serviced the time-critical interrupt.

ruuddw reviewed (commented):
I don't like the implicit priority change from 0 to 1, see inline comments

andrewboie commented on include/arch/arc/v2/irq.h:
```diff
@@ -68,6 +69,82 @@ extern void z_irq_spurious(void *unused);
+	if (IS_ENABLED(CONFIG_ARC_FIRQ) && (priority_p == 0)) { \
+		if (!IS_ENABLED(CONFIG_ARC_FIRQ_STACK) || \
+		(IS_ENABLED(CONFIG_ARC_FIRQ_STACK) && \
+		IS_ENABLED(CONFIG_ARC_STACK_CHECKING))) { \
```
Remove this if statement, both clauses are the same

vonhust replied:
fixed

andrewboie commented on arch/arc/core/fast_irq.S:
```diff
@@ -22,6 +22,23 @@
+SECTION_FUNC(TEXT, z_arc_firq_stack_set)
+	clri r1
+	mov ilink, r0
+	lr r0, [_ARC_V2_STATUS32]
```
agreed

ruuddw commented on tests/kernel/gen_isr_table/testcase.yaml:
```diff
@@ -4,3 +4,8 @@ tests:
+  arch.interrupt.arc:
+    filter: CONFIG_CPU_ARCV2
+    extra_configs:
+      - CONFIG_ARC_FIRQ_STACK=y
```
are there (other) test cases that still cover the non-FIRQ_STACK (and non-HW_STACK_PROTECTION) cases?

vonhust replied:
if ARC_FIRQ is configured and priority 0 is used, and no firq_stack, the test will fail in build because no firq_stack.
If ARC_FIRQ  is not configured, e.g nsim_sem,  or priority 0 is not used, the test will pass.

tyrantlizard77 reviewed (approved):
Looks fine. Please get CI passing. This appears to be what's killing you:

```
Error: Aborting due to non-whitelisted Kconfig warning
'/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-
out/v2m_beetle/tests/kernel/gen_isr_table/arch.interrupt.arc/sanitycheck/testcase_extra.conf:1:
warning: attempt to assign the value 'y' to the undefined symbol ARC_FIRQ_STACK'. Note: If this
warning doesn't point to an actual problem, you can add it to the whitelist at the top of
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/kconfig/kconfig.py.
```

ioannisg commented on tests/kernel/gen_isr_table/testcase.yaml:
```diff
@@ -4,3 +4,9 @@ tests:
         stm32_min_dev_blue usb_kw24d512 v2m_beetle
     filter: CONFIG_GEN_ISR_TABLES and CONFIG_ARMV7_M_ARMV8_M_MAINLINE
     tags: interrupt isr_table
+  arch.interrupt.arc:
```
why don't you add tag here?

vonhust commented:
ping for merge
